nvflare/app_common/workflows/fedopt.py

```diff
@@ -0,0 +1,166 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import time
+
+import torch
+from net import Net
+
+from nvflare.app_common.abstract.fl_model import FLModel
+from nvflare.security.logging import secure_format_exception
+
+from .fedavg import FedAvg
+
+
+class FedOpt(FedAvg):
+    def __init__(
+        self,
+        *args,
+        optimizer_args: dict = {"lr": 1.0, "momentum": 0.6},
+        lr_scheduler_args: dict = None,
+        device=None,
+        **kwargs,
+    ):
+        """Implement the FedOpt algorithm.
+
+        The algorithm is proposed in Reddi, Sashank, et al. "Adaptive federated optimization." arXiv preprint arXiv:2003.00295 (2020).
+        After each round, update the global model using the specified PyTorch optimizer and learning rate scheduler.
+        Note: This class will use FedOpt to optimize the global trainable parameters (i.e. `self.torch_model.named_parameters()`)
+        but use FedAvg to update any other layers such as batch norm statistics.
+
+        Args:
+            optimizer_args: dictionary of optimizer arguments, e.g.
+                {'lr': 1.0} (default).
+            lr_scheduler_args: dictionary of server-side learning rate scheduler arguments, e.g.
+                {'T_max': 100} (default: None).
+            device: specify the device to run server-side optimization, e.g. "cpu" or "cuda:0"
+                (will default to cuda if available and no device is specified).
+
+        Raises:
+            TypeError: when any of input arguments does not have correct type
+        """
+        super().__init__(*args, **kwargs)
+
+        if not optimizer_args:
+            self.warning("No optimizer_args provided. Using FedOpt with SGD and lr 1.0, momentum 0.6")
+            optimizer_args = {"lr": 1.0, "momentum": 0.6}
+        if not isinstance(optimizer_args, dict):
+            raise TypeError("optimizer_args must be a dict of format, e.g. {'lr': 1.0, 'momentum': 0.6}.")
+        if lr_scheduler_args is not None and not isinstance(lr_scheduler_args, dict):
+            raise TypeError("optimizer_args must be a dict of format, e.g. {'T_max': 100, 'eta_min': 0.9}")
+
+        self.optimizer_args = optimizer_args
+        self.lr_scheduler_args = lr_scheduler_args
+        if device is None:
+            self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        else:
+            self.device = torch.device(device)
+
+        self.torch_model = None
+        self.optimizer = None
+        self.lr_scheduler = None
+
+    def initialize(self, fl_ctx):
+        super().initialize(fl_ctx)
+
+        self.torch_model = Net()
+        self.torch_model.to(self.device)
+
+        # set up optimizer
+        try:
+            self.optimizer = torch.optim.SGD(
```

Comment:
Right this can be done, however we would need to use the `self.engine` `build_component()` and `get_component()`. Do we need a better way to expose these to the user in `WFController`, or is accessing `self.engine` from `ModelController` and using these functions fine? @chesterxgchen 

Also related, `self.engine` will be set in the parent’s `initialize()`, so if users override `initialize()` here (also done in Scaffold) they need to call `super().initialize()` first for the controller to work. 

One idea around this issue is renaming Controller’s `initialize()` to `initialize_controller()` to avoid overriding, and users instead only care about the `run()` routine and place these initializing logic at the beginning.

Reply:
Yes, ideally we can pass the optimizer object directly as an argument to this controller.

Reply:
In this case we can't build the optimizer beforehand as it requires the model parameters `params` as an argument for the constructor. Similarly the lr_scheduler requires the optimizer.

I can change to use `self.engine.build_component(config_dict)`, however should we make this function more visible to users via a wrapper, or leave it as is?